reactor-extra/src/main/java/reactor/retry/DefaultRetry.java

```diff
@@ -0,0 +1,150 @@
+import reactor.core.scheduler.Scheduler;
+import reactor.util.Logger;
+import reactor.util.Loggers;
+
+public class DefaultRetry<T> extends AbstractRetry<T, Throwable> implements Retry<T> {
+
+	static final Logger log = Loggers.getLogger(DefaultRetry.class);
+	static final Consumer<? super RetryContext<?>> NOOP_ON_RETRY = r -> {};
+
+	final Predicate<? super RetryContext<T>> retryPredicate;
+	final Consumer<? super RetryContext<T>> onRetry;
+
+	DefaultRetry(Predicate<? super RetryContext<T>> retryPredicate,
+			int maxIterations,
+			Duration timeout,
+			Backoff backoff,
+			Jitter jitter,
+			Scheduler backoffScheduler,
+			final Consumer<? super RetryContext<T>> onRetry,
+			T applicationContext) {
+		super(maxIterations, timeout, backoff, jitter, backoffScheduler,applicationContext);
+		this.retryPredicate = retryPredicate;
+		this.onRetry = onRetry;
+	}
+
+	public static <T> DefaultRetry<T> create(Predicate<? super RetryContext<T>> retryPredicate) {
+		return new DefaultRetry<T>(retryPredicate,
+				1,
+				null,
+				Backoff.zero(),
+				Jitter.noJitter(),
+				null,
+				NOOP_ON_RETRY,
+				(T) null);
+	}
+
+	@Override
+	public Publisher<Long> apply(Flux<Throwable> errors) {
+		return new RetryFunction().apply(errors);
+	}
+
+	@Override
+	public Retry<T> withApplicationContext(T applicationContext) {
+		return new DefaultRetry<>(retryPredicate, maxIterations, timeout,
+				backoff, jitter, backoffScheduler, onRetry, applicationContext);
+	}
+
+	@Override
+	public Retry<T> doOnRetry(Consumer<? super RetryContext<T>> onRetry) {
+		return new DefaultRetry<>(retryPredicate, maxIterations, timeout,
+				backoff, jitter, backoffScheduler, onRetry, applicationContext);
+	}
+
+	@Override
+	public Retry<T> retryMax(int maxIterations) {
+		if (maxIterations < 0)
+			throw new IllegalArgumentException("maxIterations should be >= 0");
+		return new DefaultRetry<>(retryPredicate, maxIterations, timeout,
+				backoff, jitter, backoffScheduler, onRetry, applicationContext);
+	}
+
+	@Override
+	public Retry<T> timeout(Duration timeout) {
+		if (timeout.isNegative())
+			throw new IllegalArgumentException("timeout should be >= 0");
+		return new DefaultRetry<>(retryPredicate, Integer.MAX_VALUE, timeout,
+				backoff, jitter, backoffScheduler, onRetry, applicationContext);
+	}
+
+	@Override
+	public Retry<T> backoff(Backoff backoff) {
+		return new DefaultRetry<>(retryPredicate, maxIterations, timeout,
+				backoff, jitter, backoffScheduler, onRetry, applicationContext);
+	}
+
+	@Override
+	public Retry<T> jitter(Jitter jitter) {
+		return new DefaultRetry<>(retryPredicate, maxIterations, timeout,
+				backoff, jitter, backoffScheduler, onRetry, applicationContext);
+	}
+
+	@Override
+	public Retry<T> withBackoffScheduler(Scheduler scheduler) {
+		return new DefaultRetry<>(retryPredicate, maxIterations, timeout,
+				backoff, jitter, scheduler, onRetry, applicationContext);
+	}
+
+	class RetryFunction implements Function<Flux<Throwable>, Publisher<Long>> {
```

Comment:
I wonder if there's a way of still having 1 independent `Publisher` per `DefaultRetry#apply` but still get rid of that inner `RetryFunction`. Same for `RepeatFunction`.